### PR TITLE
ARROW-1744: [Plasma] Provide TensorFlow operator to transfer Tensors between Plasma and TensorFlow

### PR DESCRIPTION
**Pull request by pschafhalter**
This is based off of #2046 from @pcmoritz and @concretevitamin, who wrote the original TensorFlow Op with help from @ericl.

In addition to the previous PR, this supports serialization of arbitrary type tensors.

To get it working, arrow has to be compiled in the following way:

```
cmake -DARROW_PLASMA=on -DARROW_TENSORFLOW=on -DARROW_PYTHON=on ..
make -j8
sudo make install
```

And pyarrow like that:

```
PYARROW_WITH_PLASMA=1 PYARROW_WITH_TENSORFLOW=1 python setup.py develop
```

The PR also includes utilities that should be generally useful in converting between TensorFlow and Arrow. It is currently supposed to be used to transfer weights in and out of TensorFlow but hopefully in the future there will also be support for data Tensors.

### CODE REVIEW

**Comment by pcmoritz** [comment]
This is ready for review

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (robertnishihara) [inline python/pyarrow/tests/test_plasma_tf_op.py]: Is this comment still true?

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (robertnishihara) [inline python/pyarrow/tests/test_plasma_tf_op.py]: check 64-byte alignment here

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (robertnishihara) [inline cpp/src/arrow/python/python_to_arrow.cc]: is this code path guaranteeing 64-byte alignment of the data buffers?

File: cpp/src/arrow/python/python_to_arrow.h
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.h]: @pitrou could you have a look at this?

File: cpp/CMakeLists.txt
Comment (wesm) [inline cpp/CMakeLists.txt]: This is set _very_ late in the game. I think you need to add this before the call to `include(ThirdpartyToolchain)` otherwise the build will be broken (dependencies built with new ABI)

File: cpp/src/arrow/adapters/tensorflow/convert.h
Comment (wesm) [inline cpp/src/arrow/adapters/tensorflow/convert.h]: `Status::TypeError`?

File: cpp/src/arrow/adapters/tensorflow/convert.h
Comment (wesm) [inline cpp/src/arrow/adapters/tensorflow/convert.h]: Do you think it would be useful to define some TF-specific metadata (to include in `custom_metadata` at the Field level) so that these types can be carried through?

File: cpp/src/arrow/adapters/tensorflow/convert.h
Comment (wesm) [inline cpp/src/arrow/adapters/tensorflow/convert.h]: Ditto here

File: python/pyarrow/tensorflow/build.sh
Comment (wesm) [inline python/pyarrow/tensorflow/build.sh]: I'm genuinely curious about why this (which at a glance looks like chicanery to me) is required versus handling this at the pyarrow (or libarrow) build stage. Would it be possible to package pyarrow with TF support out of the box?

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: Who is zongheng? I didn't see him/her in the commits

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: Is there a way to not have these global variables?

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: This reads odd, should `SequenceBuilder` have a nullary constructor?

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: Are these explicit namespaces necessary? Tensor, Buffer, and SerializedPyObject should all be visible here

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: Passing `const DataType&` here would prevent copy

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: What do you think about a namespace alias `tf` instead?

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: There is a sense of code duplication in the constructor and destructor here vs. the other kernel above. Any way to enable code reuse?

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: Should `offset` be the same as `header_size`?

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (wesm) [inline python/pyarrow/tensorflow/plasma_op.cc]: who is ekl?

**Review by wesm — COMMENTED** [review]
I left a few more comments on the C++ code. 

Can you create some follow-up JIRAs to make sure we incorporate this into our packaging scripts so this ships and works in 0.10.0?

File: cpp/CMakeLists.txt
Comment (pitrou) [reply cpp/CMakeLists.txt]: Also I'm curious about this. This depends on the TensorFlow build, right? What if someone builds TensorFlow with the "new ABI"? I would expect this kind of settings to take place in build scripts rather than be forced in the cmake files.

File: cpp/src/arrow/python/python_to_arrow.h
Comment (pcmoritz) [reply cpp/src/arrow/python/python_to_arrow.h]: This header is included in the compilation of the TensorFlow Op now, which is done after the wheels have been created, on the user's computer (we decided about this strategy in https://github.com/apache/arrow/pull/2046, see the comments from @ppwwyyxx and @xhochy). On the user's computer, we don't have a fancy build environment (no CMake, no Python paths, etc.). This change makes it possible to compile the TensorFlow Op without a Python environment.

File: cpp/CMakeLists.txt
Comment (pcmoritz) [reply cpp/CMakeLists.txt]: This is done to be compatible with the TensorFlow binaries, which are compiled against this particular ABI. I can put it earlier and can also make it configurable wrt. which ABI is used (not sure which one is used if TensorFlow is compiled from source).

There is an alternative possible solution which I have considered, which is to create a C interface between the TensorFlow op and Arrow. This would avoid having to juggle with the `_GLIBCXX_USE_CXX11_ABI` flag but makes the code significantly uglier. Any thoughts?

File: cpp/src/arrow/adapters/tensorflow/convert.h
Comment (pcmoritz) [reply cpp/src/arrow/adapters/tensorflow/convert.h]: Yes, will do!

File: cpp/src/arrow/adapters/tensorflow/convert.h
Comment (pcmoritz) [reply cpp/src/arrow/adapters/tensorflow/convert.h]: That's a good idea, we can do this as a followup PR if needed (I haven't seen the more exotic types in the wild, presumably the Q types are quantized in some way?)

File: python/pyarrow/tensorflow/build.sh
Comment (pcmoritz) [reply python/pyarrow/tensorflow/build.sh]: I did it in the pyarrow build stage in https://github.com/apache/arrow/pull/2046 but it was correctly pointed out that the TensorFlow op probably needs to be built on the user's computer to be compatible with all versions of tensorflow. This is trying to make it so we need minimal requirements (only g++ and TensorFlow headers which come with the pip package). I.e. no CMake. I'm very open to suggestions if somebody has a better way of doing this. The goal is to be compatible with as many TensorFlow versions and user configurations as possible (tensorflow-cpu, tensorflow-gpu, different versions etc) and make it robust. This will probably need some experimentation as it will be used on many different setups.

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (pcmoritz) [reply python/pyarrow/tensorflow/plasma_op.cc]: zongheng is @concretevitamin (mentioned in the PR message), he is a PhD student at the RISELab

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (pcmoritz) [reply python/pyarrow/tensorflow/plasma_op.cc]: ekl is @ericl, also a student at the RISELab who helped getting this together

File: cpp/CMakeLists.txt
Comment (pitrou) [reply cpp/CMakeLists.txt]: > This is done to be compatible with the TensorFlow binaries

I understand, but why do build-specific compilation options end up hardcoded in the cmake build files?

> There is an alternative possible solution which I have considered, which is to create a C interface between the TensorFlow op and Arrow. This would avoid having to juggle with the _GLIBCXX_USE_CXX11_ABI flag but makes the code significantly uglier. Any thoughts?

Depends on the level of ugliness and complexity I would say :-)

File: cpp/src/arrow/python/python_to_arrow.h
Comment (pitrou) [reply cpp/src/arrow/python/python_to_arrow.h]: Hmm, so my question is: why does the "TensorFlow Op" (I'm not sure what that is, sorry: is it some kind of computation kernel?) need to include the Arrow Python files?

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (pcmoritz) [reply cpp/src/arrow/python/python_to_arrow.cc]: Yeah, I'll change that (it was originally developed in the tensorflow op file and there were some name clashes and I forgot to remove the namespaces).

**Comment by pcmoritz** [comment]
Thanks for reviewing and the comments so far, I'll respond to the remaining ones and incorporate the changes tomorrow!

The most important question to figure out is the more high level question of how/when to build the op and I'm curious about more thoughts here. Building it at runtime seems inevitable to me and I would really like to avoid having to introduce a C API if possible (to avoid the ABI conflicts), but let me know what you think. Most likely there will be a need for more TensorFlow/arrow code in the future (i.e. for passing arrow datasets/RecordBatches into TensorFlow) so having a solid foundation will be valueable.

File: cpp/src/arrow/python/python_to_arrow.h
Comment (pcmoritz) [reply cpp/src/arrow/python/python_to_arrow.h]: Yeah, it's a TensorFlow extension (a shared library) that can be loaded into TensorFlow and provides custom operators, like this one which reads/writes neural network weights from and to plasma efficiently.

It needs to include the Arrow Python files, because they define a serialization format for plasma that makes it possible to read/write objects from python. We use this in Ray to communicate TensorFlow weights between python and the TensorFlow runtime in an efficient way.

Making it so python_to_arrow.h and arrow_to_python.h don't depend on Python headers might be valuable in other contexts too.

Comment (pitrou) [reply cpp/src/arrow/python/python_to_arrow.h]: Unless the serialization format is Python-specific (is it, and if so, why?), perhaps it can be defined outside of `arrow/python`.

File: cpp/CMakeLists.txt
Comment (pcmoritz) [reply cpp/CMakeLists.txt]: There is a `shared_ptr` involved that needs to be kept between various method calls, so we would at least need a `context` struct and a way to free it. Doable but not pretty. Also this code would be very specific to this operator and not as reusable as the components we have at the moment. That's why I was trying to avoid it :)

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (concretevitamin) [reply python/pyarrow/tensorflow/plasma_op.cc]: I originally authored this code in a private repo.  Then @pschafhalter and @pcmoritz took it over to productionize/push it into arrow codebase.

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (pcmoritz) [reply python/pyarrow/tests/test_plasma_tf_op.py]: Yeah!

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (pcmoritz) [reply cpp/src/arrow/python/python_to_arrow.cc]: Yeah, it uses the same code as for normal tensor serialization so will be 64-byte aligned.

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (pcmoritz) [reply python/pyarrow/tests/test_plasma_tf_op.py]: added it

File: cpp/src/arrow/adapters/tensorflow/convert.h
Comment (pcmoritz) [reply cpp/src/arrow/adapters/tensorflow/convert.h]: Done

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (robertnishihara) [inline python/pyarrow/tensorflow/plasma_op.cc]: This is one plasma client per custom op? The challenge with that is you could easily need hundreds or thousands of custom ops in a single TF graph, which would be a lot of plasma clients, potentially running out of file descriptors somewhere.

We can certainly try it this way and see what happens.

**Comment by robertnishihara** [comment]
Great work, this looks good to me! I'd hold off on introducing a C API and focus on keeping the code clean/simple for now. If we run into substantial build or binary issues, then that may be the right thing to do.

**Comment by pcmoritz** [comment]
Yeah, sounds good to me, this is ready to merge now. Let's see what our experience is deploying it with Ray and then we can iterate on it.

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (concretevitamin) [inline python/pyarrow/tensorflow/plasma_op.cc]: (Communicated offline as well, copy pasting here)

I don’t understand this fix. Why would the destructor of the same object be called twice?
Can you (1) revert this commit, (2) change the Python test from `sess = tf.Session(...); ...` to `with tf.Session(...) as sess:`?  Just want to see if this is some default graph semantics or something like that (likely not)
Another independent change to try (1) revert this commit, (2) add `.SetIsStateful()` to the op’s registration, the two REGISTER_OP() in memcpy_plasma_op.cc

**Comment by pitrou** [comment]
Regardless of the review comments, I want to ask a higher-level question: does this have to be part of the Arrow codebase? We seem to be piling up adapters for third-party libraries without regard for maintenance concerns. I think even the fact that Plasma is part of the Arrow codebase should be questioned. It clearly is a separate piece of code (to the point that it has distinct build files) with its own quality standards.

**Comment by wesm** [comment]
I think it's a good idea to have all the code in the same place precisely to ease maintenance but also to have consistent governance standards Though there are Apache projects with multiple repos, I would really like to avoid that. Even having the Parquet adapter code in a separate codebase has been more trouble than it's worth IMHO. 

The burden of maintenance should clearly be shared. I find it unfortunate that the Ray developers are not participating much in the packaging and release maintenance for Arrow, and instead ship this code on their own inside Ray.

**Comment by pitrou** [comment]
In Parquet's case I understand, because there's a weird bidirectional dependency between the Parquet and Arrow repositories that makes everything complicated. In Plasma's case, the dependency would only be in a single direction (from Plasma to Arrow). Also, Plasma is more of an application than a library that would sit at the same level as Arrow.

**Comment by wesm** [comment]
> Also, Plasma is more of an application than a library that would sit at the same level as Arrow.

This is true _for now_. We haven't built anything inside Arrow that depends on Plasma yet, but I think it's entirely possible in the future (spill-to-disk joins, etc.). In the context of a database system, Plasma serves as a "subsystem for managing shared memory pages and out-of-core data", so it fits with the objective of building a deconstructed database.

**Comment by pcmoritz** [comment]
I fully agree with @wesm on Plasma. Let me know @pitrou if you have things in mind we can improve. The codebases originally started out as separate with different styles etc., but things are converging more and more and we are happy to put in work to make it even more so in the future.

Regarding TensorFlow, my perspective on this is that machine learning is a super important part of modern data processing (increasingly so) and TensorFlow is the de facto standard for ML and therefore we need to spend the time to make arrow and TensorFlow work together very well. Even if that's going to be painful (TensorFlow has its quirks). This is going to be far easier if the work happens in the arrow repo.

@wesm Thanks for the feedback and please let us know how we can improve the cooperation on the packaging and release maintenance.

**Comment by wesm** [comment]
Here is a thread from March 23 about problems with Arrow's packaging and release management:

https://lists.apache.org/thread.html/46b7251192568258e9fc01e819fd96d103d524647c8977f1fd9046de@<dev.arrow.apache.org>

No one who develops both Ray and Arrow replied on the thread, presumably (?) because you are vendoring Arrow and dealing with packaging on your own, so the woes that the Arrow release managers have experienced haven't impacted you very much. There's nothing inherently wrong or immoral with vendoring an upstream open source project. 

In principle, you could base Ray releases on released versions of Arrow (Ray has released about as many times as Arrow has). To make this practical, Arrow would need to release more often. BUT, Arrow has not released more often because the release management and packaging process has been so historically onerous (I have borderline PTSD from having been the release manager too many times, only half joking).

Meanwhile, our 0.10.0 release is blocked as we've decided not to release again unless we can include the binary packages in the release vote, to prevent the kind of packaging disaster that befell the 0.9.0 release. So this is where we stand right now.

Since we are all ultimately volunteers, there's nothing forcing you to help with packaging and releases for Arrow, but the community would be stronger and healthier if you did.

**Comment by wesm** [comment]
BTW, Ray's vendored pyarrow conflicts with the non-vendored one if they are both installed:

```
>>> import ray
>>> import pyarrow
>>> pyarrow
<module 'pyarrow' from '/home/wesm/miniconda/envs/ray-test/lib/python3.6/site-packages/ray/pyarrow_files/pyarrow/__init__.py'>
```

I took a look at the PyPI download stats on BigQuery -- pyarrow is being installed ~10-15x as often as Ray, so this could spell trouble for users.

**Comment by pcmoritz** [comment]
We do want to switch over to including pyarrow via pip into ray; we can try looking into that. The requirement from the development side is to have arrow nightlies or per-commit packages that we can use for the development of Ray (since there is a lot of development on both sides that needs to be in sync). For releases we would need to synchronize with arrow releases in some way.

The advantage to vendor arrow with Ray like we do now is that there is just one version we need to support and from the user's perspective "it just works".

**Review by pcmoritz — APPROVED** [review]

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (robertnishihara) [reply python/pyarrow/tests/test_plasma_tf_op.py]: Can you make a JIRA for fixing this?

File: python/pyarrow/compat.py
Comment (robertnishihara) [inline python/pyarrow/compat.py]: in the case where this fails, why not try `import tensorflow`?

Comment (pcmoritz) [reply python/pyarrow/compat.py]: It happens inside of the manylinux1 build (the OSError occurs if the the version of GLIBC is too old for TensorFlow); in this case import tensorflow also fails.

Comment (robertnishihara) [reply python/pyarrow/compat.py]: Ok, I see.

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (robertnishihara) [reply python/pyarrow/tests/test_plasma_tf_op.py]: If you compose the write op with the read op, don't we want that to be the identity?

File: python/pyarrow/tensorflow/plasma_op.cc
Comment (robertnishihara) [reply python/pyarrow/tensorflow/plasma_op.cc]: What do you mean there is one op per graph. Couldn't there be an arbitrary number of ops per graph?

Comment (pcmoritz) [reply python/pyarrow/tensorflow/plasma_op.cc]: In principle yes, but one per graph is how we are using it, I think it's fine to have one plasma client per op (and it's cleaner code wise since we don't need global variables in this case)

File: python/pyarrow/tests/test_plasma_tf_op.py
Comment (pcmoritz) [reply python/pyarrow/tests/test_plasma_tf_op.py]: That requires to output a list of Tensors (with lengh determine at runtime). I created a JIRA here: https://issues.apache.org/jira/browse/ARROW-2866

**Comment by wesm** [comment]
Thanks all! If any of you have a moment, maybe you can jot down some ideas for where to go with TF integration from here? https://cwiki.apache.org/confluence/display/ARROW#ApacheArrowHome-MachineLearningFrameworkIntegrations

**Comment by yupbank** [comment]
this should be relevant https://github.com/tensorflow/io/tree/master/tensorflow_io/arrow 

https://github.com/tensorflow/io/pull/36#issuecomment-456223528

**Comment by xhochy** [comment]
I would be really happen when this code could move to tensorflow/io. It has caused quite some headaches in our packaging scripts.

**Pull request closed without merging**
